pkg/nmi/server/redirector_linux.go

```diff
@@ -0,0 +1,83 @@
+// +build linux
+
+package server
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	iptables "github.com/Azure/aad-pod-identity/pkg/nmi/iptables"
+	"k8s.io/klog"
+)
+
+// LinuxRedirector returns sync function for linux redirector
+func LinuxRedirector(server *Server) func(*Server, chan bool, chan bool) {
+	return func(server *Server, subRoutinedone chan bool, mainRoutineDone chan bool) {
+		updateIPTableRules(server, subRoutinedone, mainRoutineDone)
+	}
+}
+
+// WindowsRedirector returns sync function for windows redirector
+func WindowsRedirector(server *Server) func(*Server, chan bool, chan bool) {
+	panic("Windows Redirector is not applicable")
+}
+
+func updateIPTableRulesInternal(server *Server) {
+	klog.V(5).Infof("node(%s) hostip(%s) metadataaddress(%s:%s) nmiport(%s)", server.NodeName, server.HostIP, server.MetadataIP, server.MetadataPort, server.NMIPort)
+
+	if err := iptables.AddCustomChain(server.MetadataIP, server.MetadataPort, server.HostIP, server.NMIPort); err != nil {
+		klog.Fatalf("%s", err)
+	}
+	if err := iptables.LogCustomChain(); err != nil {
+		klog.Fatalf("%s", err)
+	}
+}
+
+// updateIPTableRules ensures the correct iptable rules are set
+// such that metadata requests are received by nmi assigned port
+// NOT originating from HostIP destined to metadata endpoint are
+// routed to NMI endpoint
+func updateIPTableRules(server *Server, subRoutinedone chan bool, mainRoutineDone chan bool) {
```

Comment:
Channels could be Read, Write and R/W. Channel owner is the code that has Write permissions to the channel - it is the one who can close the channel. Common pattern in Go is to limit channel only to one of those modes which clearly communicates to the developers if this part of the code is supposed to only read, write to the channel or both. Passing channel with write permission also indicates the transfer of the ownership of the channel - the code that passes a write channel should not attempt to close it.

Reply:
Add the directions to the channel.

Reply:
Good point!